url270fc4f5ee_ykcilborw_Joust_tgz-pJ8-fri_patterns_interpreter_parsergenerator_examples_HowToOverrideLexerImplJ8/src/fri/patterns/interpreter/parsergenerator/semantics/ReflectSemantic.java

```diff
@@ -0,0 +1,121 @@
+package fri.patterns.interpreter.parsergenerator.semantics;
+
+import java.lang.reflect.*;
+import java.util.*;
+import fri.patterns.interpreter.parsergenerator.syntax.Rule;
+import fri.patterns.interpreter.parsergenerator.Semantic;
+
+/**
+	Semantic base class that tries to call methods with the same name as nonterminal
+	on left side of rule. The number of parameters of the method must be the same as
+	the number of symbols on the right side of the rule, all parameters must be of type
+	Object, and the return type must be Object. The method might process the parse
+	results and return an arbitrary Object to be processed by another callback method.
+	The ranges of the currently dispatched rule is available by protected getRanges() method.
+	<p/>
+	Example: Both rules
+	<pre>
+		expression ::= expression "+" term ;
+		expression ::= expression "-" term ;
+	</pre>
+	would be dispatched by
+	<pre>
+		public Object expression(Object expression, Object operator, Object term)	{
+			int e = ((Integer) expression).intValue();
+			int t = ((Integer) term).intValue();
+			if (operator.equals("+"))
+				return new Integer(e + t);
+			else
+			if (operator.equals("-"))
+				return new Integer(e - t);
+			else
+				throw new IllegalArgumentException("Unknown operator: "+operator);
+		}
+	</pre>
+	
+	@see fri.patterns.interpreter.parsergenerator.examples.Calculator
+	@author (c) 2000, Fritz Ritzberger
+*/
+
+public abstract class ReflectSemantic implements Semantic
+{
+	private List ranges;
+	
+	/**
+		Tries to find a method with same name as left side of rule and number of arguments
+		as much as elements in inputTokens list. All argument types are Object.
+		@param rule the rule that was recognized by the parser.
+		@param inputTokens all semantic call returns from underlying rules, collected according to current rule.
+		@return object to push on parser value-stack.
+	*/
+	public Object doSemantic(Rule rule, List inputTokens, List ranges)	{
+		this.ranges = ranges;
+		
+		String nonterminal = rule.getNonterminal();
+		
+		Class [] types = new Class [inputTokens.size()];
+		for (int i = 0; i < types.length; i++)
+			types[i] = Object.class;
+		
+		Method m = null;
+		try	{
+			m = getClass().getMethod(nonterminal, types);
+		}
+		catch (Exception e)	{
+			//System.err.println("WARNING: method not implemented: "+nonterminal+", input is "+inputTokens);
+			return fallback(rule, inputTokens, ranges);
+		}
+		
+		Object [] args = new Object[inputTokens.size()];
+		inputTokens.toArray(args);
+
+		if (m != null)	{		
+			try	{
+				m.setAccessible(true);	// problem with anonymous inner classes
+				Object o = m.invoke(this, args);
+				return o;
+			}
+			catch (IllegalAccessException e)	{
+				e.printStackTrace();
+			}
+			catch (IllegalArgumentException e)	{
+				e.printStackTrace();
+			}
+			catch (InvocationTargetException e)	{
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)